vcd/resource_vcd_dnat.go

```diff
@@ -96,31 +101,64 @@ func resourceVcdDNATCreate(d *schema.ResourceData, meta interface{}) error {
 		icmpSubType = ""
 	}
 
+	var orgVdcnetwork *types.OrgVDCNetwork
+	providedNetworkName := d.Get("network_name")
+	if nil != providedNetworkName && "" != providedNetworkName {
+		orgVdcnetwork, err = getNetwork(d, vcdClient, providedNetworkName.(string))
+	} else {
+		_, _ = fmt.Fprint(GetTerraformStdout(), "WARNING: This resource will require network_name in the next major version \n")
+	}
+	if err != nil {
+		return fmt.Errorf("unable to find orgVdcnetwork: %s, err: %s", providedNetworkName.(string), err)
+	}
+
 	// Creating a loop to offer further protection from the edge gateway erroring
 	// due to being busy eg another person is using another client so wouldn't be
 	// constrained by out lock. If the edge gateway reurns with a busy error, wait
 	// 3 seconds and then try again. Continue until a non-busy error or success
 
-	err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
-		task, err := edgeGateway.AddNATPortMapping("DNAT",
-			d.Get("external_ip").(string),
-			portString,
-			d.Get("internal_ip").(string),
-			translatedPortString, protocol,
-			icmpSubType)
-		if err != nil {
-			return resource.RetryableError(
-				fmt.Errorf("error setting DNAT rules: %#v", err))
-		}
+	if nil != providedNetworkName && providedNetworkName != "" {
+		err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
+			task, err := edgeGateway.AddNATPortMappingWithUplink(orgVdcnetwork, "DNAT",
+				d.Get("external_ip").(string),
+				portString,
+				d.Get("internal_ip").(string),
+				translatedPortString, protocol,
+				icmpSubType)
+			if err != nil {
+				return resource.RetryableError(
+					fmt.Errorf("error setting DNAT rules: %#v", err))
+			}
 
-		return resource.RetryableError(task.WaitTaskCompletion())
-	})
+			return resource.RetryableError(task.WaitTaskCompletion())
+		})
+	} else {
+		// TODO remove when major release is done
+		err = retryCall(vcdClient.MaxRetryTimeout, func() *resource.RetryError {
+			task, err := edgeGateway.AddNATPortMapping("DNAT",
+				d.Get("external_ip").(string),
+				portString,
+				d.Get("internal_ip").(string),
+				translatedPortString, protocol,
+				icmpSubType)
+			if err != nil {
+				return resource.RetryableError(
+					fmt.Errorf("error setting DNAT rules: %#v", err))
+			}
+
+			return resource.RetryableError(task.WaitTaskCompletion())
+		})
+	}
 
 	if err != nil {
 		return fmt.Errorf("error completing tasks: %#v", err)
 	}
 
-	d.SetId(d.Get("external_ip").(string) + ":" + portString + " > " + d.Get("internal_ip").(string) + ":" + translatedPortString)
+	if nil != providedNetworkName && "" != providedNetworkName {
```

Comment:
It hold until 3.0, but real IDs are something like:
<img width="636" alt="image" src="https://user-images.githubusercontent.com/15804230/56583444-3e13a580-65e2-11e9-9177-e5d51ef40da2.png">

Then it is easy to do lookups.

Reply:
It's common approach we have now , that resources have their own Id's - not Vcd's. I think option to add to TODO's to analyse that.

Reply:
Yeah. I don't mind being like that now. IMHO it's just for future and lookups would be way more convenient than with current ID in this case.

Reply:
I agree